File: java/solutions/03-advanced/ex071_lru_cache/LruCache.java
package fewolearning.exercises.advanced.ex071_lru_cache;

import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.Map;

/*
Exercise 071 - LRU cache (reference solution).
*/
public final class LruCache<K, V> {
    private final int capacity;
    private final Map<K, V> store;

    public LruCache(int capacity) {
        this.capacity = capacity;
        this.store = new LinkedHashMap<>(16, 0.75f, true);
    }

    public void put(K key, V value) {
        store.put(key, value);
        if (store.size() > capacity) {
            Iterator<Map.Entry<K, V>> eldestFirst = store.entrySet().iterator();
            eldestFirst.next();
            eldestFirst.remove();
        }
    }

    public V get(K key) {
        return store.get(key);
    }

    public int size() {
        return store.size();
    }
}

File: php/solutions/03-advanced/ex071_lru_cache/LruCache.php
<?php

declare(strict_types=1);

namespace FeWoLearning\Exercises\Advanced\Ex071LruCache;

final class LruCache
{
    private array $store = [];

    public function __construct(private readonly int $capacity)
    {
    }

    public function get(string $key): mixed
    {
        if (!array_key_exists($key, $this->store)) {
            return null;
        }

        $value = $this->store[$key];
        unset($this->store[$key]);
        $this->store[$key] = $value;

        return $value;
    }

    public function put(string $key, mixed $value): void
    {
        if (array_key_exists($key, $this->store)) {
            unset($this->store[$key]);
        }

        $this->store[$key] = $value;

        if (count($this->store) > $this->capacity) {
            $oldestKey = array_key_first($this->store);
            unset($this->store[$oldestKey]);
        }
    }
}
